R/bundlePackageRenv.R

```diff
@@ -0,0 +1,139 @@
+
+    deps$description <- lapply(deps$Package, function(pkg) {
+      dcf <- read.dcf(file.path(lib_dir, pkg, "DESCRIPTION"))
+      as.list(as.data.frame(dcf))
+    })
+  }
+
+  deps
+}
+
+standardizeRenvPackages <- function(packages, repos, biocPackages = NULL) {
+  repos <- standardizeRepos(repos)
+  availablePackages <- availablePackages(repos)
+
+  names(packages) <- NULL
+  out <- lapply(
+    packages,
+    standardizeRenvPackage,
+    availablePackages = availablePackages,
+    biocPackages = biocPackages,
+    repos = repos
+  )
+  out <- compact(out)
+  out <- lapply(out, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(out)
+}
+
+standardizeRenvPackage <- function(pkg,
+                                   availablePackages,
+                                   biocPackages = NULL,
+                                   repos = character(),
+                                   bioc) {
+  # Don't include renv itself
+  if (identical(pkg$Package, "renv")) {
+    return(NULL)
+  }
+
+  # Convert renv source to manifest source/repository
+  # https://github.com/rstudio/renv/blob/0.17.2/R/snapshot.R#L730-L773
+
+  if (is.null(pkg$Repository) && !is.null(pkg$RemoteRepos) && grepl("bioconductor.org", pkg$RemoteRepos)) {
+    # Work around bug where renv fails to detect BioC package installed by pak
+    # https://github.com/rstudio/renv/issues/1202
+    pkg$Source <- "Bioconductor"
+  }
+
+  if (pkg$Source == "Repository") {
+    if (pkg$Repository == "CRAN") {
+      if (isDevVersion(pkg, availablePackages)) {
+        pkg$Source <- NA_character_
+        pkg$Repository <- NA_character_
+      } else {
+        pkg$Source <- "CRAN"
+        pkg$Repository <- findRepoUrl(pkg$Package, availablePackages)
+      }
+    } else {
+      # $Repository comes from DESCRIPTION and is set by repo, so can be
+      # anything. So we must look up from the package name
+      pkg$Repository <- findRepoUrl(pkg$Package, availablePackages)
+      pkg$Source <- findRepoName(pkg$Repository, repos)
+    }
+  } else if (pkg$Source == "Bioconductor") {
+    pkg$Repository <- findRepoUrl(pkg$Package, availablePackages)
+    if (is.na(pkg$Repository)) {
+      # Try packages defined from default bioC repos
+      pkg$Repository <- findRepoUrl(pkg$Package, biocPackages)
+    }
+  } else if (pkg$Source == "unknown") {
+    pkg$Source <- NA_character_
+  } else if (pkg$Source %in% c("BitBucket", "GitHub", "GitLab")) {
+    pkg$Source <- tolower(pkg$Source)
+  }
+
+  # Remove Remote fields that pak adds for "standard" installs from CRAN
+  if (identical(pkg$RemoteType, "standard")) {
+    pkg <- pkg[!grepl("^Remote", names(pkg))]
+  }
+
+  pkg[manifestPackageColumns(pkg)]
+}
+
+biocPackages <- function(bundleDir) {
+  signal("evaluating", class = "rsconnect_biocRepos") # used for testing
+  availablePackages(biocRepos(bundleDir))
+}
+biocRepos <- function(bundleDir) {
+  repos <- getFromNamespace("renv_bioconductor_repos", "renv")(bundleDir)
```

Comment:
@kevinushey do we also need to consider `options(BioC_mirror)` or does `renv_bioconductor_repos()` already do that?

Reply:
`renv` basically just calls `BiocManager:::.repositories()` so that that should respect BioC_mirror IIUC.

Reply:
I'll probably switch this to use the `git_branch` `DESCRIPTION` (with some error/fallback) as it sounds like it's ok to rely on that: https://github.com/Bioconductor/BiocManager/issues/164#issuecomment-1519088086